presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +79,332 @@
+     * as well as guarding against the above situation, so it's all around
+     * better.
+     *
+     * The other, other approach would be to set an 'invalidate after' time
+     * in the holder. Now we have to worry about clock resolution and
+     * monotonicity. No thanks; somebody did that work for the ScheduledFuture/
+     * ScheduledExecutorService, let's take advantage of that.
+     *
+     * The locking hierarchy is as follows:
+     * 1) The monitor on the ReferenceCountingLoadingCache
+     * 2) <internal locking of com.google.common.cache.LoadingCache>
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount by calling the cache's
+     * close() method, and in doing so closes their HTTP clients. If anybody is
+     * still trying to use a PrestoConnection created by the driver, they're
+     * SOL.
+     */
+    private static class ReferenceCountingLoadingCache<K, V>
+    {
+        private class Holder
+        {
+            private final V value;
+            private int refcount;
+            private int cleanerCount;
+            private ScheduledFuture oldCleanup;
+
+            private Holder(V value)
+            {
+                this.value = requireNonNull(value);
+            }
+
+            private V get()
+            {
+                return value;
+            }
+
+            private int getRefcount()
+            {
+                return refcount;
+            }
+
+            private int getCleanerCount()
+            {
+                return cleanerCount;
+            }
+
+            private int reference()
+            {
+                setCleanup(null);
+                return ++refcount;
+            }
+
+            private int dereference()
+            {
+                return --refcount;
+            }
+
+            private void setCleanup(ScheduledFuture newCleanup)
+            {
+                if (oldCleanup != null && oldCleanup.cancel(false)) {
+                    --cleanerCount;
+                }
+                oldCleanup = newCleanup;
+            }
+
+            private void scheduleCleanup(
+                    ScheduledExecutorService cleanupService,
+                    Runnable cleanupTask,
+                    long delay,
+                    TimeUnit unit)
+            {
+                checkState(refcount == 0, "non-zero refcount in scheduleCleanup");
+                ++cleanerCount;
+                ScheduledFuture newCleanup = cleanupService.schedule(cleanupTask, delay, unit);
+                setCleanup(newCleanup);
+            }
+
+            private void releaseForCleaning()
+            {
+                --cleanerCount;
+            }
+        }
+
+        private final ScheduledExecutorService valueCleanupService;
+        private final LoadingCache<K, Holder> backingCache;
```

Comment:
We don't, but we get all of the actual cache management logic for free, leaving the code here to deal only with the synchronization bits. I'm not married to this, but keeping it separate seems more clear. Also, there's no contention for the loading cache's synchronization since it's already synchronized here.